libzmix/src/signatures/delg_cred_cdd/groth_sig.rs

```diff
@@ -0,0 +1,439 @@
+use super::errors::{DelgError, DelgResult};
+use amcl_wrapper::extension_field_gt::GT;
+use amcl_wrapper::field_elem::{FieldElement, FieldElementVector};
+use amcl_wrapper::group_elem::{GroupElement, GroupElementVector};
+use amcl_wrapper::group_elem_g1::{G1LookupTable, G1Vector, G1};
+use amcl_wrapper::group_elem_g2::{G2Vector, G2};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct GrothSigkey(pub FieldElement);
+
+macro_rules! impl_GrothS {
+    ( $GrothSetupParams:ident, $GrothVerkey:ident, $GrothSig:ident, $GrothS:ident, $vk_group:ident, $msg_group:ident, $GVector:ident ) => {
+        #[derive(Clone, Debug, Serialize, Deserialize)]
+        pub struct $GrothSetupParams {
+            pub g1: G1,
+            pub g2: G2,
+            pub y: $GVector,
+        }
+
+        #[derive(Clone, Debug, Serialize, Deserialize)]
+        pub struct $GrothVerkey(pub $vk_group);
+
+        #[derive(Clone, Debug, Serialize, Deserialize)]
+        pub struct $GrothSig {
+            pub R: $vk_group,
+            pub S: $msg_group,
+            pub T: $GVector,
+        }
+
+        pub struct $GrothS {}
+    };
+}
+
+macro_rules! impl_GrothS_setup {
+    ( $GrothSetupParams:ident, $msg_group:ident, $GVector:ident ) => {
+        pub fn setup(count_messages: usize, label: &[u8]) -> $GrothSetupParams {
```

Comment:
Is the label necessary for security?

Reply:
To guarantee that generators come from random oracle